packages/jaeger-ui/src/utils/DeepDependencyGraph/parse-payload.test.js

```diff
@@ -0,0 +1,86 @@
+// Copyright (c) 2019 Uber Technologies, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import _filter from 'lodash/filter';
+import _flatten from 'lodash/flatten';
+import _map from 'lodash/map';
+
+import parsePayload from './parse-payload';
+import * as testResources from './parse-payload.test.resources';
+
+describe('parse payload', () => {
+  function parsedOutputValidator({ paths: payload, focalIndices, groupOperations = false }) {
+    const { focalPathElem } = testResources;
+    const focalPathElemArgument = groupOperations ? { service: focalPathElem.service } : focalPathElem;
+    const { paths, services } = parsePayload(payload, focalPathElemArgument);
+    const serviceNames = Object.keys(services);
+
+    expect(serviceNames).toEqual(Array.from(new Set(_map(_flatten(payload), 'service'))));
+    serviceNames.forEach(serviceName => {
+      expect(Object.keys(services[serviceName].operations)).toEqual(
+        Array.from(new Set(_map(_filter(_flatten(payload), { service: serviceName }), 'operation')))
+      );
+    });
```

Comment:
Why not use native JS functionality for this type of thing?

```js
// if `services` and `service.operations` are changed to `Map`s
const inputFlat = [].concat(...payload);

const inputSvcs = new Set(inputFlat.map(item => item.service));
expect(new Set(services.keys())).toEqual(inputSvcs);

for (let [name, service] of services) {
  const ops = inputFlat.map(item => item.service === service ? item.operation : null).filter(Boolean);
  expect(new Set(service.operations.keys())).toEqual(new Set(ops));
}
```

Reply:
To me, the value of using lodash here is that it is very declarative. The five nested function calls is dense, but I can remove `Array.from` and add `new Set` to the `expect` part. Then it reads as `_flatten` the payload, `_filter` to those with the target `serviceName`, `_map` to get `operation`, then create the `new Set` for comparison.

The code you provided is good, but I have to read more carefully to validate it. IMO `const ops = inputFlat.map(item => item.service === service ? item.operation : null).filter(Boolean);` is as dense as `_map(_filter(_flatten(payload), { service: serviceName }), 'operation')` and requires returning `null` from the `map` in order to do the `.filter(Boolean)` which isn't bad but is more to remember while checking the code.

Reply:
I think it's about preference. Personally, I lean toward native. But, to each their own.